src/arch/wasm/CodeGen.zig

```diff
@@ -4990,19 +4990,33 @@ fn fpext(func: *CodeGen, operand: WValue, given: Type, wanted: Type) InnerError!
     const given_bits = given.floatBits(func.target);
     const wanted_bits = wanted.floatBits(func.target);
 
-    if (wanted_bits == 64 and given_bits == 32) {
-        try func.emitWValue(operand);
+    var current_operand = operand;
+    var current_bits = given_bits;
+    var current_ty = given;
+
+    if (wanted_bits == 64 and given_bits == 16) {
+        current_operand = try func.callIntrinsic("__extendhfsf2", &.{given}, Type.f32, &.{current_operand});
+        current_bits = 32;
+        current_ty = Type.f32;
+    }
+
+    if (wanted_bits > 32 and current_bits == 32) {
+        try func.emitWValue(current_operand);
         try func.addTag(.f64_promote_f32);
-        return .stack;
+        current_operand = .stack;
+        current_bits = 64;
+        current_ty = Type.f64;
```

Comment:
I find this pattern quite error-prone and hard to read. I think keeping the very original code with only the `TODO` branch updated to make use of the dynamic compiler-rt call keeps it much more clear. e.g:
```zig
fn fpext(func: *CodeGen, operand: WValue, given: Type, wanted: Type) InnerError!WValue {
    const given_bits = given.floatBits(func.target);
    const wanted_bits = wanted.floatBits(func.target);

    if (wanted_bits == 64 and given_bits == 32) {
        try func.emitWValue(operand);
        try func.addTag(.f64_promote_f32);
        return WValue{ .stack = {} };
    } else if (given_bits == 16) {
        // call __extendhfsf2(f16) f32
        const f32_result = try func.callIntrinsic(
            "__extendhfsf2",
            &.{Type.f16},
            Type.f32,
            &.{operand},
        );

        if (wanted_bits == 32) {
            return f32_result;
        }
        if (wanted_bits == 64) {
            try func.addTag(.f64_promote_f32);
            return WValue{ .stack = {} };
        }
    }

     // put the dynamic call to compiler-rt here
}
```

Repeating some code to keep it clear is always a winner compared to trying to be clever in my books.